packages/core-bridge/src/runtime.rs

```diff
@@ -1,557 +1,446 @@
-
-/// Convert Rust SystemTime into a JS array with 2 numbers (seconds, nanos)
-pub fn system_time_to_js<'a, C>(cx: &mut C, time: SystemTime) -> NeonResult<Handle<'a, JsArray>>
-where
-    C: Context<'a>,
-{
-    let nanos = time
-        .duration_since(UNIX_EPOCH)
-        .unwrap_or(Duration::ZERO)
-        .as_nanos();
-    let only_nanos = cx.number((nanos % 1_000_000_000) as f64);
-    let ts_seconds = cx.number((nanos / 1_000_000_000) as f64);
-    let ts = cx.empty_array();
-    ts.set(cx, 0, ts_seconds).unwrap();
-    ts.set(cx, 1, only_nanos).unwrap();
-    Ok(ts)
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[macro_export]
+macro_rules! enter_sync {
+    ($runtime:expr) => {
+        let _trace_subscriber_guard = $runtime
+            .telemetry()
+            .trace_subscriber()
+            .map(|sub| tracing::subscriber::set_default(sub));
+        let _tokio_handle_guard = $runtime.tokio_handle().enter();
+    };
 }
 
-/// Helper to get the current time in nanosecond resolution.
-pub fn get_time_of_day(mut cx: FunctionContext) -> JsResult<JsArray> {
-    system_time_to_js(&mut cx, SystemTime::now())
+pub trait RuntimeExt {
+    fn future_to_promise<F, R>(&self, future: F) -> BridgeResult<BridgeFuture<R>>
+    where
+        F: Future<Output = Result<R, BridgeError>> + Send + 'static,
+        R: TryIntoJs + Send + 'static;
 }
 
-/// Initialize Core global telemetry and create the tokio runtime required to run Core.
-/// This should typically be called once on process startup.
-/// Immediately spawns a poller thread that will block on [RuntimeRequest]s
-pub fn runtime_new(mut cx: FunctionContext) -> JsResult<BoxedRuntime> {
-    let telemetry_options = cx.argument::<JsObject>(0)?.as_telemetry_options(&mut cx)?;
-    let channel = Arc::new(cx.channel());
-    let (sender, mut receiver) = unbounded_channel::<RuntimeRequest>();
-
-    // FIXME: This is a temporary fix to get sync notifications of errors while initializing the runtime.
-    //        The proper fix would be to avoid spawning a new thread here, so that start_bridge_loop
-    //        can simply yeild back a Result. But early attempts to do just that caused panics
-    //        on runtime shutdown, so let's use this hack until we can dig deeper.
-    let (result_sender, result_receiver) = oneshot::channel::<Result<(), String>>();
-
-    std::thread::spawn(move || {
-        start_bridge_loop(telemetry_options, channel, &mut receiver, result_sender)
-    });
-
-    if let Ok(Err(e)) = result_receiver.blocking_recv() {
-        Err(cx.throw_error::<_, String>(e).unwrap_err())?;
+impl RuntimeExt for CoreRuntime {
+    fn future_to_promise<F, R>(&self, future: F) -> BridgeResult<BridgeFuture<R>>
+    where
+        F: Future<Output = Result<R, BridgeError>> + Send + 'static,
+        R: TryIntoJs + Send + 'static,
+    {
+        enter_sync!(self);
+        Ok(BridgeFuture::new(Box::pin(future)))
```

Comment:
I think this would work better with `.instrument()`? I'm surprised this works since the subscriber guard would be dropped immediately after this function exits

Reply:
The future will execute on one of Tokio’s thread, which are bound with specific tracing contexts immediately after being created.

AFAICT the node thread would not need to access the tracing subscriber once the implementation of the “js_function” returns.